docs_src/design/adr/security/0015-in-cluster-tls.md

```diff
@@ -0,0 +1,229 @@
+  that the requests are being routed to the correct server.
+
+- Server authentication of clients.
+  The server knows the identity of the client that has connected to it.
+  There are a variety of mechanims to achieve this,
+  such as usernames and passwords, tokens, claims, et cetera,
+  but the mechanism under consideration by this ADR
+  is TLS client authentication
+  using TLS client certificates.
+
+- Confidentiality of messages exchanged between services.
+  Confidentiality is needed to protect authentication data
+  flowing between communicating microservices as well as
+  to protect the message payloads if they contain nonpublic data.
+  TLS provides communication channel confidentiality.
+
+- Integrity of messages exchanged between services.
+  Integrity is needed to ensure that messages between
+  communicating microservices are not maliciously altered,
+  such as inserting or deleting data in the middle of the exchange.
+  TLS provides communication channel integrity.
+
+A microservice architecture normally strives for all of the above protections.
+
+Besides TLS, there are other mechanisms that can be used to provide some of the above properties.
+For example, IPSec tunnels provide confidentity, integrity, and authentication of the hosts (network-level protection).
+SSH tunnels provide confidentiality, integrity, and authentication of the tunnel endpoints (also network-level protection).
+TLS, however, is preferred, because it operates in-process at the application level and provides better point-to-point security.
+
+### Why to not encrypt?
+
+In the case of TLS communications,
+microservices depend on an asymmetric private key to prove their identity.
+To be of value, this private key must be kept secret.
+Applications typically depend on process-level isolation
+and/or file system protections for the private key.
+Moreover, interprocess communication using sockets is mediated by the
+operating system kernel.
+An attacker running at the privilege of the operating system
+has the ability to compromise TLS protections,
+such as by substituting a private key or certificate authority of their choice,
+accessing the unencrypted data in process memory,
+or intercepting the network communications that flow through the kernel.
+Therefore, within a single host, TLS protections may slow down an attacker,
+but are not likely to stop them.
+Additionally, use of TLS requires management of additional
+security assets in the form of TLS private keys.
+
+Microservice communcation across hosts, however,
+is vulnerable to intereception, and must be protected
+via some mechanism such as, but not limited to:
+IPSec or SSH tunnels, encrypted overlay networks,
+service mesh middlewares, or application-level TLS.
+
+
+## Decision
+
+At this time, EdgeX is primarily a single-node IoT application framework.
+Should this position change, this ADR should be revisited.
+Based on the single-node assumption:
```

Comment:
IMHO, can also mention the reduced overheads of using non-TLS vs. TLS internally like cipher algo.exchange, certificate handshaking, ssl handshaking overhead, .. what have you.